contracts/factory/src/contract.rs

```diff
@@ -518,6 +528,86 @@ impl FactoryTrait for Factory {
 
         Ok(())
     }
+
+    fn propose_admin(
+        env: Env,
+        new_admin: Address,
+        time_limit: Option<u64>,
+    ) -> Result<Address, ContractError> {
+        let current_admin = get_config(&env).admin;
+        current_admin.require_auth();
+
+        if current_admin == new_admin {
+            log!(&env, "Trying to set new admin as new");
+            panic_with_error!(&env, ContractError::SameAdmin);
+        }
+
+        env.storage().instance().set(
+            &PENDING_ADMIN,
+            &AdminChange {
+                new_admin: new_admin.clone(),
+                time_limit,
+            },
+        );
+
+        env.events().publish(
+            ("Factory: ", "Admin replacement requested by old admin: "),
+            &current_admin,
+        );
+        env.events()
+            .publish(("Factory: ", "Replace with new admin: "), &new_admin);
+
+        Ok(new_admin)
+    }
+
+    fn revoke_admin_change(env: Env) -> Result<(), ContractError> {
+        let current_admin = get_config(&env).admin;
+        current_admin.require_auth();
+
+        if !env.storage().instance().has(&PENDING_ADMIN) {
+            log!(&env, "No admin change in place");
+            panic_with_error!(&env, ContractError::NoAdminChangeInPlace);
+        }
+
+        env.storage().instance().remove(&PENDING_ADMIN);
+
+        env.events()
+            .publish(("Factory: ", "Undo admin change: "), ());
```

Comment:
nit: Those events could be grouped together and pushed at once with events `all()`: https://docs.rs/soroban-sdk/latest/soroban_sdk/testutils/trait.Events.html#tymethod.all